contrib/smtp_proxy/filters/network/source/smtp_decoder_impl.cc

```diff
@@ -0,0 +1,112 @@
+#include "contrib/smtp_proxy/filters/network/source/smtp_decoder_impl.h"
+
+#include "source/common/common/logger.h"
+#include "source/extensions/filters/network/well_known_names.h"
+
+#include "absl/strings/match.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace SmtpProxy {
+
+DecoderImpl::DecoderImpl(DecoderCallbacks* callbacks, TimeSource& time_source,
+                         Random::RandomGenerator& random_generator)
+    : callbacks_(callbacks), time_source_(time_source), random_generator_(random_generator) {
+  session_ = new SmtpSession(callbacks, time_source_, random_generator_);
+}
+
+SmtpUtils::Result DecoderImpl::onData(Buffer::Instance& data, bool upstream) {
+  SmtpUtils::Result result = SmtpUtils::Result::ReadyForNext;
+  if (upstream) {
+    result = parseResponse(data);
+    data.drain(data.length());
+    return result;
+  }
+  result = parseCommand(data);
+  data.drain(data.length());
+  return result;
+}
+
+SmtpUtils::Result DecoderImpl::parseCommand(Buffer::Instance& data) {
+  ENVOY_LOG(debug, "smtp_proxy parseCommand: decoding {} bytes", data.length());
+
+  SmtpUtils::Result result = SmtpUtils::Result::ReadyForNext;
+
+  if (getSession()->isTerminated()) {
+    return result;
+  }
+  if (getSession()->isDataTransferInProgress()) {
+    getSession()->updateBytesMeterOnCommand(data);
+    return result;
+  }
+
+  std::string buffer = data.toString();
+  ENVOY_LOG(debug, "received command {}", buffer);
+  // Each SMTP command ends with CRLF ("\r\n"), if received buffer doesn't end with CRLF, the filter
+  // will not process it.
+  if (!absl::EndsWith(buffer, SmtpUtils::smtpCrlfSuffix)) {
+    return result;
+  }
+
+  buffer = StringUtil::cropRight(buffer, SmtpUtils::smtpCrlfSuffix);
+  int length = buffer.length();
+
+  std::string command = "";
+  std::string args = "";
+  if (length < 4) {
```

Comment:
I don't think this is going to be able to transparently handle unknown commands since it could be something like BDAT or AUTH/SASL modes (other than PLAIN) that take the protocol out of "line mode".

I would just compare to a list of the commands you know how to handle and disable protocol inspection if you see anything else.